### ISSUE TITLE
Exception when fetching data using CARBONLINK_HOSTS

### ISSUE THREAD

**Issue by loopping** [graphite-project/graphite-web#1139]
Hello,
I'm using latest master graphite-web, carbon, whisper to date and django 1.4
When I set the CARBONLINK_HOST to a carbon-cache daemon (127.0.0.1:7002:a), I have this error when trying to get a "render".
If I comment CARBONLINK_HOST, no more python errors but I lose the benefit of the 
cache from carbon and I have delay in the graphs.

Has anyone the same error ?
Regards

<pre><code>Fri Feb 06 13:56:48 2015 :: Failed after 2 retry! See: 'float' object is not iterable
Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 174, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
Fri Feb 06 13:56:48 2015 :: Exception encountered in <POST http://myhost:8080/render>
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/django/core/handlers/base.py", line 111, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/opt/graphite/webapp/graphite/render/views.py", line 111, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 7, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 25, in evaluateTokens
    args = [evaluateTokens(requestContext, arg) for arg in tokens.call.args]
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 25, in evaluateTokens
    args = [evaluateTokens(requestContext, arg) for arg in tokens.call.args]
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 18, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 21, in evaluateTokens
    return fetchData(requestContext, tokens.pathExpression)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 141, in fetchData
    raise Exception("Failed after %i retry! See: %s" % (settings.MAX_FETCH_RETRIES, e))
Exception: Failed after 2 retry! See: 'float' object is not iterable

</code></pre>

**Comment by loopping** [comment]
Hello, more info on my errors:

When using a carbonlink and with this http query the graph is ok:
http://myurl:8080/render/?target=nmon.os1.cpu_all.Idle_perc

If I try this http query I get the error mentionned above:
http://myurl:8080/render/?target=nmon.os1.cpu_all.Idle_perc&target=nmon.os1.Sys_perc

Regards

**Comment by loopping** [comment]
Hello some more news about carbonlink and cache query:
I have set some log debug in the file /opt/graphite/webapp/graphite/carbonlink.py in the function.
In the class  CarbonLinkPool and in the function send_request (line 106), I printed request and result:
Here is what I got:
<pre><code>send_request:request:{'metric': 'nmon.os1.cpu_all.Idle_perc', 'type': 'cache-query'}
send_request:result:{'datapoints': {1423491741.0: 2.2000000000000002, 1423491681.0: 3.1000000000000001, 1423491621.0: 2.2000000000000002, 1423491801.0: 2.1000000000000001}}
</code></pre>
I was wondering if the error I have "float' object is not iterable" might come from the epoch time has a ".0" in it.

**Comment by afilipchik** [comment]
I see same error with the latest build.

**Comment by afilipchik** [comment]
the hacky fix is to add:
```python
 if hasattr(cached_datapoints, 'iteritems'):
        cached_datapoints = cached_datapoints.iteritems()
```
to
```python
    try:
      cached_datapoints = CarbonLink.query(self.real_metric_path)
    except:
      log.exception("Failed CarbonLink query '%s'" % self.real_metric_path)
      cached_datapoints = {}

    if hasattr(cached_datapoints, 'iteritems'):
      cached_datapoints = cached_datapoints.iteritems()

    for (timestamp, value) in cached_datapoints
      interval = timestamp - (timestamp % data.timeStep)

      try:
        i = int(interval - data.startTime) / data.timeStep
        values[i] = value
      except:
        pass
```

**Comment by brutasse** [comment]
In carbon's master branch, cached datapoints are stored as a dict whereas in the 0.9.x branch they're a list-like structure (deque IIRC).$

This is how I account for the two possible return values in graphite-api: https://github.com/brutasse/graphite-api/blob/master/graphite_api/finders/whisper.py#L122-L123

**Comment by loopping** [comment]
Some more logs on this subject:
I logged the content of datapoints,result in function stringReceived in file lib/carbon/protocols.py.
I logged the content of request,results in function query in file webapp/graphite/carbonlink.py
I logged the content of cached_datapoints in function fetch in file webapp/graphite/readers.py
I have edited the carbon/examples/example-client.py, changed DELAY from 60 to 0.
I ran example-client.py for a few seconds and asked graphite to render carbon.agents.vmlinux-a.metricsReceived and refreshed every 5 sec or so.
Here are the contents of datapoints,request and results when I have the error mentionned above:

<pre><code>
Sun Feb 22 09:34:48 2015 :: carbonlink:query:request:{'metric': 'carbon.agents.vmlinux-a.metricsReceived', 'type': 'cache-query'}
None
22/02/2015 10:34:48 :: protocols:stringReceived:datapoints for carbon.agents.vmlinux-a.metricsReceived:{1424597616.247383: 822, 1424597676.270806: 848, 1424597556.274894: 852}:
22/02/2015 10:34:48 :: protocols:stringReceived:result for carbon.agents.vmlinux-a.metricsReceived:{'datapoints': {1424597616.247383: 822, 1424597676.270806: 848, 1424597556.274894: 852}}
Sun Feb 22 09:34:48 2015 :: carbonlink:query:results:{'datapoints': [1424597616.247383, 1424597556.274894, 1424597676.270806]}
None
Sun Feb 22 09:34:48 2015 :: readers:fetch:cached_datapoints:[1424597616.247383, 1424597556.274894, 1424597676.270806]
None
Sun Feb 22 09:34:48 2015 :: Got an exception when fetching data! See: 'float' object is not iterable Will do it again! Run: 1 of 2
Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 177, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
</code></pre>

In some conditions I can't find, values are not fetched correctly from the cache
